bindings/experimental/distrdf/python/DistRDF/HeadNode.py

```diff
@@ -0,0 +1,377 @@
+        """
+        Python 2 dunder method for __bool__. Kept for compatibility.
+        """
+        return self.__bool__()
+
+
+class Factory(object):
+    """
+    A factory for different kinds of root nodes of the RDataFrame computation
+    graph, depending on the arguments to the RDataFrame constructor. Currently
+    can create a TreeHeadNode or an EntriesHeadNode.
+    """
+
+    @staticmethod
+    def get_headnode(*args):
+        """
+        Parses the arguments and compares them against the possible RDataFrame
+        constructors
+        """
+
+        # Early check that arguments are accepted by RDataFrame
+        ROOT.RDataFrame(*args)
+
+        firstarg = args[0]
+        if isinstance(firstarg, int):
+            # RDataFrame(ULong64_t numEntries)
+            return EntriesHeadNode(firstarg)
+        elif isinstance(firstarg, (ROOT.TTree, str)):
+            # RDataFrame(std::string_view treeName, filenameglob, defaultBranches = {})
+            # RDataFrame(std::string_view treename, filenames, defaultBranches = {})
+            # RDataFrame(std::string_view treeName, dirPtr, defaultBranches = {})
+            # RDataFrame(TTree &tree, const ColumnNames_t &defaultBranches = {})
+            return TreeHeadNode(*args)
+        else:
+            raise RuntimeError(
+                ("First argument {} of type {} is not recognised as a supported "
+                 "argument for distributed RDataFrame. Currently only TTree/Tchain "
+                 "based datasets or datasets created from a number of entries "
+                 "can be processed distributedly.").format(firstarg, type(firstarg)))
+
+
+class EntriesHeadNode(Node.Node):
+    """
+    The head node of a computation graph where the RDataFrame is created from
+    a number of entries.
+    """
+
+    def __init__(self, nentries):
+        """
+        Creates a new RDataFrame instance for the given arguments.
+
+        Args:
+            nentries (int): The number of entries this RDataFrame will process.
+        """
+        super(EntriesHeadNode, self).__init__(None, None)
+
+        self.nentries = nentries
+        # Set at creation of the dataframe, might be optimized by the backend
+        # in optimize_partitions
+        self.npartitions = 2
+
+    def build_ranges(self):
+        """Build the ranges for this dataset."""
+        # Empty datasets cannot be processed distributedly
+        if not self.nentries:
+            raise RuntimeError(
+                ("Cannot build a distributed RDataFrame with zero entries. "
+                 "Distributed computation will fail. "))
+        if self.npartitions > self.nentries:
+            # Restrict 'npartitions' if it's greater than 'nentries'
+            msg = ("Number of partitions {0} is greater than number of entries {1} "
+                   "in the dataframe. Using {1} partition(s)".format(self.npartitions, self.nentries))
+            warnings.warn(msg, UserWarning, stacklevel=2)
```

Comment:
Yes they get a warning. I'm not sure it's valid usage. There is already a case of "valid" local usage and invalid distributed usage in DistRDF, namely when a user supplies a TTree with 0 entries. That's invalid because there is no point in running a distributed execution on 0 entries and it will finally lead to crashes on the distributed workers which are hard to deal with (difficult to output user friendly warnings or errors from there). So we already have a `RuntimeError` in place for that case.

This case is slightly different, I think it's not strictly an error wanting to run on 1 entry, but at least users should know they should provide enough data to create the minimum number of partitions. Maybe the warning could be improved for this particular case.

Reply:
being discussed in the context of https://github.com/root-project/root/pull/8485